pysat/utils/io.py

```diff
@@ -716,9 +738,15 @@ def load_netcdf_xarray(fnames, strict_meta=False, file_format='NETCDF4',
         data = xr.open_mfdataset(fnames, decode_timedelta=decode_timedelta,
                                  combine='by_coords')
 
-    # TODO(#947) Add conversion for timestamps that may have been treated
-    # incorrectly, including origin and unit.  At the moment this is only done
-    # for Pandas data.
+    # If epoch can be converted to a datetime, convert
+    try:
+        data[epoch_name] = xr.DataArray(pds.to_datetime(data[epoch_name],
+                                                        unit=epoch_unit,
+                                                        origin=epoch_origin),
+                                        coords=data[epoch_name].coords)
+    except (KeyError, ValueError):
+        # Value does not exist or cannot be converted to datetime
+        pass
```

Comment:
Open design question: What happens if epoch data is non-convertable?
- Option A (used currently): pass and send info to the logger
- Option B: Check that data is a float / int and exists rather than use try / except.  If data is empty or already a datetime, this conversion routine is not run.  

The key difference is that sometimes a float / int that lies outside the datetime range for given epoch info will be passed through (for instance, if the user specifies a bad epoch or unit for conversion).  Do we maintain the error in this case, or leave the data unconverted and log the error as info?  Currently, this will just skip the conversion.

@aburrell, @rstoneback, any thoughts on the preferred behaviour here?

Reply:
- First thought is the KeyError condition for epoch_name should raise an error.
 - For the ValueError. If the errors that come up are fixable by the user, like specifying the correct units, etc, then I think we should error. Make the user correct the data. 
 - If there are non-fixable things, like perhaps the file is weird, then I'm favor of using logging.warning().

Reply:
I agree with @rstoneback

Reply:
It looks like this is a bug related to the "Santa does not exist" issue above.  `inst_to_netcdf` does not create an epoch if a random one is specified for xarray objects.  This creates the `KeyError` since this data is not contained in the datafile.  This was not discovered before since there was no check that the epoch exists in current tests.  However adding an automated calculation based on epoch having a value would error.  The "except KeyError" case was covering up a deeper bug in the system.

I propose creating an out of scope issue for this bug.  Instead of a try/except structure, check if the epoch exists, then skip the math and raise a warning to the user.  Any errors are still raised if the epoch exists and is weird.  Something like:
```
    if epoch_name in data.variables:
        data[epoch_name] = xr.DataArray(pds.to_datetime(data[epoch_name],
                                                        unit=epoch_unit,
                                                        origin=epoch_origin),
                                        coords=data[epoch_name].coords)
    else:
        pysat.logger.warning('{:} not found in datafile'.format(epoch_name))
```

The original intent of the try / except statement for ValueErrors was to skip over the conversion if the data was already a datetime object, since datasets in pysat libraries may convert separately (eg, ICON data).  However, those are done as preprocess routines, which means they are run after the load.  Additionally, they use the general library, which is updated in this pull to skip when the data is already in a datetime format.  Which means that the "except ValueError" case is not needed here.

Reply:
As far as the unit tests go, it creates 11 errors.  I haven't seen any issues in the other packages, but could conceivably create an issue if you save an ND dataset with a custom epoch name.

```
====================================== short test summary info =======================================
FAILED pysat/tests/test_utils_io.py::TestLoadNetCDFXArray::test_basic_write_and_read_netcdf_mixed_case_data_format
FAILED pysat/tests/test_utils_io.py::TestLoadNetCDFXArray::test_basic_write_and_read_netcdf_mixed_case_meta_format
FAILED pysat/tests/test_utils_io.py::TestLoadNetCDFXArray::test_inst_write_and_read_netcdf - KeyErr...
FAILED pysat/tests/test_utils_io.py::TestLoadNetCDFXArray::test_write_and_read_netcdf4_w_kwargs[wkwargs0-lkwargs0]
FAILED pysat/tests/test_utils_io.py::TestLoadNetCDFXArray::test_write_and_read_netcdf4_w_kwargs[wkwargs1-lkwargs1]
FAILED pysat/tests/test_utils_io.py::TestLoadNetCDFXArray::test_write_and_read_netcdf4_w_kwargs[wkwargs2-lkwargs2]
FAILED pysat/tests/test_utils_io.py::TestLoadNetCDFXArray::test_write_and_read_netcdf4_w_kwargs[wkwargs3-lkwargs3]
FAILED pysat/tests/test_utils_io.py::TestLoadNetCDFXArray::test_netcdf_attribute_override - KeyErro...
FAILED pysat/tests/test_utils_io.py::TestLoadNetCDFXArray::test_read_netcdf4_with_time_meta_labels[kwargs0-False]
FAILED pysat/tests/test_utils_io.py::TestLoadNetCDFXArray::test_read_netcdf4_with_time_meta_labels[kwargs1-False]
FAILED pysat/tests/test_utils_io.py::TestLoadNetCDFXArray::test_read_netcdf4_with_time_meta_labels[kwargs2-True]
```

Reply:
It does look like the epoch issue is unrelated to your direct code changes here. Making an issue is a reasonable way to go. I'd like to get the bug sorted out before the next release.

Reply:
I'm going through details of file writing over in #983 so my plan has been to work the epoch bug in after that pull, or perhaps part of.